internal/charmstore/resources.go

```diff
@@ -0,0 +1,256 @@
+	blobName, err := s.storeResource(blob)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	doc := &mongodoc.Resource{
+		BaseURL:    entity.BaseURL,
+		Name:       name,
+		Revision:   newRevision,
+		BlobHash:   blob.Hash,
+		Size:       blob.Size,
+		BlobName:   blobName,
+		UploadTime: time.Now().UTC(),
+	}
+	if s.insertResource(doc); err != nil {
+		if err := s.BlobStore.Remove(blobName); err != nil {
+			logger.Errorf("cannot remove blob %s after error: %v", blobName, err)
+		}
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s *Store) insertResource(doc *mongodoc.Resource) error {
+	if err := doc.Validate(); err != nil {
+		return errgo.Mask(err)
+	}
+	err := s.DB.Resources().Insert(doc)
+	if err != nil {
+		if mgo.IsDup(err) {
+			return errgo.WithCausef(nil, params.ErrDuplicateUpload, "")
+		}
+		return errgo.Notef(err, "cannot insert resource")
+	}
+	return nil
+}
+
+func (s *Store) storeResource(blob ResourceBlob) (string, error) {
+	name := bson.NewObjectId().Hex()
+	// TODO(ericsnow) We will finish this in a follow-up patch.
+	return name, nil
+}
+
+func (s *Store) setResource(entity *mongodoc.Entity, channel params.Channel, name string, revision int) error {
+	if channel == params.NoChannel {
+		return errgo.New("missing channel")
+	}
+	if channel == params.UnpublishedChannel {
+		return errgo.New("cannot publish to unpublished channel")
+	}
+
+	doc, err := s.resource(entity.BaseURL, name, revision)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	if !charmHasResource(entity.CharmMeta, doc.Name) {
+		return errgo.Mask(err)
+	}
+
+	update := bson.D{{fmt.Sprintf("channelresources.%s.%s", channel, name), revision}}
+	if err := s.updateBaseEntity(entity.BaseURL, bson.D{{"$set", update}}); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s *Store) latestResource(baseEntity *mongodoc.BaseEntity, channel params.Channel, name string) (*mongodoc.Resource, error) {
+	revision, err := s.latestResourceRevision(baseEntity, channel, name)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+	doc, err := s.resource(baseEntity.URL, name, revision)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+	return doc, nil
+}
+
+func (s *Store) latestResourceRevision(baseEntity *mongodoc.BaseEntity, channel params.Channel, name string) (int, error) {
+	if channel == params.UnpublishedChannel {
+		var res mongodoc.Resource
+		err := s.DB.Resources().Find(bson.D{{"baseurl", baseEntity.URL}, {"name", name}}).Sort("-revision").One(&res)
+		if err == mgo.ErrNotFound {
+			return -1, resourceNotFound
+		}
+		if err != nil {
+			return -1, errgo.Mask(err)
+		}
+		return res.Revision, nil
+	}
+	revision, ok := baseEntity.ChannelResources[channel][name]
+	if !ok {
+		return -1, resourceNotFound
+	}
+	return revision, nil
+}
+
+func (s *Store) resource(url *charm.URL, name string, revision int) (*mongodoc.Resource, error) {
```

Comment:
// resource returns the resource with the given URL, name and revision.
// If the resource was not found, it returns an error with a resourceNotFound
// cause.